### PR TITLE
Bug 1770677: operator/ingress: add dnsrecord type

### PR DESCRIPTION
**Pull request by ironcladlou**
This commit moves the DNS CRD used by the ingress operator[1] into
the public API package as part of the overal effort to centralize
these types for consistency/visibility/maintenance.

There are no material changes to the type, only doc improvements.

[1] https://github.com/openshift/cluster-ingress-operator/tree/master/pkg/api/v1

### CODE REVIEW

**Comment by ironcladlou** [comment]
@deads2k @damemi @Miciah @danehans PTAL. When I apply this to an existing cluster, the CRD contents [look correct](https://gist.github.com/ironcladlou/2c7cba3f9e0392e256e88a39564aa357), but `oc explain` still omits a description:

```
oc explain dnsrecords --api-version=ingress.operator.openshift.io/v1
KIND:     DNSRecord
VERSION:  ingress.operator.openshift.io/v1
DESCRIPTION:
     <empty>
```

Not sure yet what's up.

**Comment by ironcladlou** [comment]
And from `oc get --raw /openapi/v2`:

```
"io.openshift.operator.ingress.v1.DNSRecord": {
      "type": "object",
      "x-kubernetes-group-version-kind": [
        {
          "group": "ingress.operator.openshift.io",
          "kind": "DNSRecord",
          "version": "v1"
        }
      ]
    },
    "io.openshift.operator.ingress.v1.DNSRecordList": {
      "description": "DNSRecordList is a list of DNSRecord",
      "required": [
        "items"
      ],
      "properties": {
...
```

Which seems out of sync with the CRD. I'm probably missing something silly...

File: operator/ingress/v1/types.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the dnsecord.
+	Spec DNSRecordSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the dnsRecord.
+	Status DNSRecordStatus `json:"status,omitempty"`
+}
+
+// dnsRecordSpec are the details of a DNS record.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: `dnsRecordSpec` → `DNSRecordSpec` (only field names should be lower-cased).  "are" → "is".

File: operator/ingress/v1/types.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+type DNSRecordSpec struct {
+	// dnsName is the hostname of the DNS record
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSName string `json:"dnsName,omitempty"`
+	// targets are DNS resource record targets.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: Omit "resource"?

File: operator/ingress/v1/types.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Targets []string `json:"targets,omitempty"`
+	// recordType is a resource record type.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: "resource" → "DNS"?  "a" → "the"?

File: operator/ingress/v1/types.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	//
+	// The currently supported values are A and CNAME.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: Maybe use double-quotes?  Also, "supported" is a loaded word, and which record type can be used depends on the platform; maybe it would be better to rephrase this as, 'For example, "A" or "CNAME".'

File: operator/ingress/v1/types.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	RecordType DNSRecordType `json:"recordType,omitempty"`
+	// ttl is the resource record TTL in seconds. The default is 30.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: "ttl" → "recordTTL".

File: operator/ingress/v1/types.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	RecordTTL int64 `json:"recordTTL,omitempty"`
+}
+
+// dnsRecordStatus is the more recently observed status of the record.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: `dnsRecordStatus` → `DNSRecordStatus`.  "more" → "most".  "the record" → "each record".

File: operator/ingress/v1/types.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+type DNSRecordStatus struct {
+	// zones are the status of the record per zone.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: Maybe "per zone" → "in each zone".

File: operator/ingress/v1/types.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Zones []DNSZoneStatus `json:"zones,omitempty"`
+}
+
+// dnsZoneStatus is the status of a record within a specific zone.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: `dnsZoneStatus` → `DNSZoneStatus`.

File: operator/ingress/v1/types.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+type DNSZoneStatus struct {
+	// dnsZone is the zone where the record is expected to be managed.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: Omit "expected to be"?  The thing that publishes status is the thing that manages the record.  Alternatively, maybe change "managed" → "published" because the thing that manages the record may have failed to publish it.

File: operator/ingress/v1/types.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	// +optional
+	Conditions []DNSZoneCondition `json:"conditions,omitempty"`
+}
+
+var (
+	// Failed means the record is not available within a zone.
+	DNSRecordFailedConditionType = "Failed"
+)
+
+// dnsZoneCondition is just the standard condition fields.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: `dnsZoneCondition` → `DNSZoneCondition`.

**Comment by ironcladlou** [comment]
Fixed. The problem was because the CRD needed `preserveUnknownFields: false` in the spec.

File: operator/ingress/v1/types.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+	CNAMERecordType string = "CNAME"
+
+	// CNAME is an RFC 1035 A record.
+	ARecordType string = "A"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+
+// dnsRecordList contains a list of dnsrecords.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: `dnsRecordList` → `DNSRecordList`.

**Comment by ironcladlou** [comment]
CRD diff against https://github.com/openshift/cluster-ingress-operator/blob/master/manifests/00-custom-resource-definition-internal.yaml:

```diff
1,2d0
< 
< ---
15a14
>   preserveUnknownFields: false
20c19,23
<       description: DNSRecord represents a DNS record.
---
>       description: "DNSRecord is a DNS record managed in the zones defined by dns.config.openshift.io/cluster
>         .spec.publicZone and .spec.privateZone. \n Cluster admin manipulation of this
>         resource is not supported. This resource is only for internal communication
>         of OpenShift operators."
>       type: object
34a38,43
>           description: spec is the specification of the desired behavior of the dnsecord.
>           type: object
>           required:
>           - dnsName
>           - recordType
>           - targets
37c46
<               description: The hostname of the DNS record
---
>               description: dnsName is the hostname of the DNS record
40,41c49,50
<               description: TTL for the record
<               format: int64
---
>               description: ttl is the resource record TTL in seconds. The default
>                 is 30.
42a52
>               format: int64
44c54,55
<               description: RecordType type of record, e.g. CNAME, A, SRV, TXT etc
---
>               description: "recordType is a resource record type. \n The currently
>                 supported values are A and CNAME."
47c58,59
<               description: The targets the DNS record points to
---
>               description: targets are DNS resource record targets.
>               type: array
50,51d61
<               type: array
<           type: object
52a63,64
>           description: status is the most recently observed status of the dnsRecord.
>           type: object
54a67,68
>               description: zones are the status of the record per zone.
>               type: array
55a70,74
>                 description: dnsZoneStatus is the status of a record within a specific
>                   zone.
>                 type: object
>                 required:
>                 - dnsZone
57a77,79
>                     description: conditions are any conditions associated with the
>                       record in the zone.
>                     type: array
58a81,83
>                       description: dnsZoneCondition is just the standard condition
>                         fields.
>                       type: object
61d85
<                           format: date-time
62a87
>                           format: date-time
71,75d95
<                       required:
<                       - status
<                       - type
<                       type: object
<                     type: array
77,78c97,99
<                     description: DNSZone is used to define a DNS hosted zone. A zone
<                       can be identified by an ID or tags.
---
>                     description: dnsZone is the zone where the record is expected
>                       to be managed.
>                     type: object
90,91d110
<                         additionalProperties:
<                           type: string
96,102c115,116
<                     type: object
<                 required:
<                 - dnsZone
<                 type: object
<               type: array
<           type: object
<       type: object
---
>                         additionalProperties:
>                           type: string
```

File: operator/ingress/install.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package operator
```
Comment (Miciah) [inline operator/ingress/install.go]: Should this be `package ingress`?

**Comment by ironcladlou** [comment]
This is ready for review.

File: operator/ingress/v1/types.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// dnsZoneStatus is the status of a record within a specific zone.
+type DNSZoneStatus struct {
+	// dnsZone is the zone where the record is expected to be managed.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSZone configv1.DNSZone `json:"dnsZone"`
+	// conditions are any conditions associated with the record in the zone.
```
Comment (Miciah) [inline operator/ingress/v1/types.go]: Probably should mention the possible conditions: 'If publishing the record fails, the "Failed" condition will be set with a reason and message describing the cause of the failure.'

**Comment by ironcladlou** [comment]
@Miciah thanks for the review, all feedback incorporated

File: operator/ingress/v1/types.go (hunk 10 above)
Comment (Miciah) [reply operator/ingress/v1/types.go]: Hold on, I thought godoc for types didn't go in the CRD, but this one _is_ in the CRD description—did CRD generation change?

Comment (Miciah) [reply operator/ingress/v1/types.go]: @damemi, is this a defect in the CRD generation?  It makes no sense to use the type name in this context: 
 https://github.com/openshift/api/blob/41c658872c442d2a9218fe9fa5e4885c260cc3ff/operator/ingress/v1/0000_50_dns-record.yaml#L70-L71

Comment (damemi) [reply operator/ingress/v1/types.go]: Godoc for types does go in the CRD, it's where descriptions have always come from

Comment (Miciah) [reply operator/ingress/v1/types.go]: Hm, I see the same thing in other CRDs with lists of objects that have godoc (for example, I see the same issue with `spec.servers` in the `dnses.operator` CRD).

To be clear, the problem is that we define a type with godoc:
```go
// Foo is ...
type Foo {
  // ...
```
We define some type that we use in a CRD and that has a field for which the type of a slice of the first type:
```go
type Bar {
  // baz is a list of ...
  baz []Foo `json:"baz"`
  // ...
```
And the generated CRD description uses the first type's godoc, something like the following:
```yaml
spec:
  validation:
    openAPIV3Schema:
      type: object
      properties:
        type: object
        baz:
          description: baz is a list of ...
          type: array
          items:
            description: Foo is ...
```
In the context of the CRD, the type name is meaningless.

Assuming the CRD generation is working as intended, we should change our godoc to omit the type name if the type is used in a slice type for a field that is used in a CRD (in particular, for `dnsrecords`, that's `DNSZoneStatus` and `DNSZoneCondition`); agreed?

So we should change "dnsZoneCondition is just the standard condition fields." to "The standard condition fields." and "DNSZoneStatus is the status of a record within a specific zone." to "The status of a record within a specific zone."

Comment (damemi) [reply operator/ingress/v1/types.go]: Isn't the convention with godoc to start a description with the type name? So that outcome might just be a requirement

Maybe instead, our comments should be more descriptive of the types they are using. Following your example, the doc could be `bas is a list of Foo`. So the user asks, what is Foo? And the next description tells them `Foo is ...`

Comment (Miciah) [reply operator/ingress/v1/types.go]: > Isn't the convention with godoc to start a description with the type name? So that outcome might just be a requirement

We already eschew that convention for fields.

> Maybe instead, our comments should be more descriptive of the types they are using. Following your example, the doc could be `bas is a list of Foo`. So the user asks, what is Foo? And the next description tells them `Foo is ...`

That would be weird and verbose but at least the result would be coherent.

Comment (ironcladlou) [reply operator/ingress/v1/types.go]: I think I got lost as to what further changes you guys want me to make. I'd like to get this merged today.

Comment (ironcladlou) [reply operator/ingress/v1/types.go]: The answer to all of these godoc questions seems unambiguous to me: the Go types should adhere to the godoc conventions, full stop. If the generator pipeline needs to do weird stuff with the casing for OpenAPI generation, whatever, that stuff shouldn't be bleeding into the godoc in a way that causes us to diverge from the godoc standards.

While I disagree with bending the Go convention to suit some specific generator pipeline's weird deficiencies or impedance mismatches with Go, I guess it's a much bigger discussion.

**Comment by Miciah** [comment]
We can sort out godoc for fields that have slice types in a follow-up.
/lgtm

File: hack/lib/init.sh [hunk 14]
```diff
@@ -23,6 +23,7 @@ network/v1 \
 oauth/v1 \
 openshiftcontrolplane/v1 \
 operator/v1 \
+operator/ingress/v1 \
```
Comment (soltysh) [inline hack/lib/init.sh]: Is there any particular reason you're introducing a new package here, instead of just using `operator/v1` like all the others?

Comment (ironcladlou) [reply hack/lib/init.sh]: The new type is in a different group, and mixing the group within the same package was causing me trouble. Maybe it's possible and I just didn't know the right incantations. If we _can_ mix groups within a single package, I can go back and try to figure it out. WDYT? cc @sttts

Comment (soltysh) [reply hack/lib/init.sh]: @ironcladlou it was more of a question why you need that new group, mixing groups is definitely not recommended approach.

Comment (ironcladlou) [reply hack/lib/init.sh]: Because we're moving an existing API group/type from one repo (cluster-ingress-operator) to this one. Using the sub-group was in hindsight probably a mistake, but now it's out there. The goal is to get the type into the API repo and out of the operator repo, not to change the API itself. Make sense?

Comment (ironcladlou) [reply hack/lib/init.sh]: @soltysh @sttts @damemi Let me know how I should proceed here. Either this is ready to merge or I need to restructure in a way that satisfies everyone but preserves our `ingress.operator.openshift.io` group.

Comment (sttts) [reply hack/lib/init.sh]: Don't use operator/ingress, but operatoringress. This flat directory structure is a must.

**Comment by sttts** [comment]
/hold

until directory structure is fixed.

File: operator/ingress/install.go [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+	Install = schemeBuilder.AddToScheme
+)
+
+func Resource(resource string) schema.GroupResource {
+	return schema.GroupResource{Group: GroupName, Resource: resource}
+}
+
+func Kind(kind string) schema.GroupKind {
+	return schema.GroupKind{Group: GroupName, Kind: kind}
+}
```
Comment (sttts) [inline operator/ingress/install.go]: this is not the pattern of all the other `register.go` files. Please follow what the other groups do.

File: operator/ingress/v1/0000_50_dns-record.yaml [hunk 16]
```diff
@@ -0,0 +1,131 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (sttts) [inline operator/ingress/v1/0000_50_dns-record.yaml]: use v1

File: operator/ingress/install.go (hunk 15 above)
Comment (sttts) [reply operator/ingress/install.go]: oops, sorry, was reading that this is in the v1 package. Nvmd.

File: operator/ingress/v1/types.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+// Cluster admin manipulation of this resource is not supported. This resource
+// is only for internal communication of OpenShift operators.
+type DNSRecord struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the dnsecord.
+	Spec DNSRecordSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the dnsRecord.
+	Status DNSRecordStatus `json:"status,omitempty"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: omitempty on non-pointers has no effect.

File: operator/ingress/v1/types.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+// dns.config.openshift.io/cluster .spec.publicZone and .spec.privateZone.
+//
+// Cluster admin manipulation of this resource is not supported. This resource
+// is only for internal communication of OpenShift operators.
+type DNSRecord struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the dnsecord.
+	Spec DNSRecordSpec `json:"spec,omitempty"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: isn't spec required?

File: operator/ingress/v1/types.go [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+	Status DNSRecordStatus `json:"status,omitempty"`
+}
+
+// DNSRecordSpec contains the details of a DNS record.
+type DNSRecordSpec struct {
+	// dnsName is the hostname of the DNS record
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSName string `json:"dnsName,omitempty"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: required and omitempty are suspicious. I would expect that empty strings are disallowed.

File: operator/ingress/v1/types.go [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+	// targets are record targets.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Targets []string `json:"targets,omitempty"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: similar to above: are empty slices allowed?

File: operator/ingress/v1/types.go [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+	// recordType is the DNS record type. For example, "A" or "CNAME".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	RecordType DNSRecordType `json:"recordType,omitempty"`
+	// recordTTL is the record TTL in seconds. The default is 30.
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	RecordTTL int64 `json:"recordTTL,omitempty"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: is zero allowed?

File: operator/ingress/v1/types.go [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+
+var (
+	// Failed means the record is not available within a zone.
+	DNSRecordFailedConditionType = "Failed"
+)
+
+// DNSZoneCondition is just the standard condition fields.
+type DNSZoneCondition struct {
+	Type               string      `json:"type"`
+	Status             string      `json:"status"`
```
Comment (sttts) [inline operator/ingress/v1/types.go]: both required

File: operator/ingress/v1/types.go (hunk 21 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: No, and it was never a pointer, so I'm going to make it required. Is there an openapi validation to enforce >=0?

Comment (damemi) [reply operator/ingress/v1/types.go]: Looking at `controller-gen crd -ww`, I think you can use this marker:
```
+kubebuilder:validation:MinItems field
	=<int>  specifies the minimun length for this list.
```

File: operator/ingress/v1/doc.go [hunk 23]
```diff
@@ -0,0 +1,7 @@
+// +k8s:deepcopy-gen=package,register
+// +k8s:defaulter-gen=TypeMeta
+// +k8s:openapi-gen=true
+
+// +kubebuilder:validation:Optional
```
Comment (damemi) [inline operator/ingress/v1/doc.go]: btw because you have this in your doc.go, it sets every field in the package to optional by default, so you don't need any of the `+kubebuilder:validation:Optional` or `+optional` comments for CRD generation. Not a big deal but just letting you know

File: hack/lib/init.sh (hunk 14 above)
Comment (ironcladlou) [reply hack/lib/init.sh]: Fixed, thanks

File: operator/ingress/v1/0000_50_dns-record.yaml (hunk 16 above)
Comment (ironcladlou) [reply operator/ingress/v1/0000_50_dns-record.yaml]: When I changed this to v1, the CRD generator silently stops working for the file. Not sure why. I notice there are many other CRDs using v1beta1.... any ideas?

File: operator/ingress/v1/types.go (hunk 18 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: Yes, fixed

File: operator/ingress/v1/types.go (hunk 19 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: Agree, fixed here and throughout (I think)

File: operator/ingress/v1/types.go (hunk 21 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: Fixed and added minimum validation

File: operator/ingress/v1/types.go (hunk 22 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: Fixed

File: operator/ingress/v1/doc.go (hunk 23 above)
Comment (ironcladlou) [reply operator/ingress/v1/doc.go]: Yeah — I actually would almost rather explicitly mark every single field with optional or required. The mostly invisible defaulting from the package doc file seems like a recipe for confusion and errors

**Comment by ironcladlou** [comment]
@sttts 
Thanks for the thorough review. All feedback addressed except for the v1beta1 issue (https://github.com/openshift/api/pull/584#pullrequestreview-358115921). 

In any case I'd like to see a successful CI run of https://github.com/openshift/cluster-ingress-operator/pull/360 before merging.

File: operatoringress/v1/types.go [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+
+var (
+	// Failed means the record is not available within a zone.
+	DNSRecordFailedConditionType = "Failed"
+)
+
+// DNSZoneCondition is just the standard condition fields.
+type DNSZoneCondition struct {
+	// +kubebuilder:validation:Required
+	// +required
```
Comment (sttts) [inline operatoringress/v1/types.go]: does this really work on the type? Would only expect those annotations on the field instances. This feels more like a property within a struct.

File: operatoringress/v1/types.go [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+	Status DNSRecordStatus `json:"status"`
+}
+
+// DNSRecordSpec contains the details of a DNS record.
+type DNSRecordSpec struct {
+	// dnsName is the hostname of the DNS record
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	DNSName string `json:"dnsName"`
```
Comment (sttts) [inline operatoringress/v1/types.go]: this will allow and empty string. Intentional?

Comment (sttts) [reply operatoringress/v1/types.go]: Similarly for targets.

Comment (ironcladlou) [reply operatoringress/v1/types.go]: Really? I thought required would cover that.... do I have to use MinLength?

File: operator/ingress/v1/types.go (hunk 20 above)
Comment (ironcladlou) [reply operator/ingress/v1/types.go]: Your note about required on a string field has me wondering whether I need something in addition to required here?

**Comment by ironcladlou** [comment]
@sttts after talking with @deads2k, my confusion was that I was not thinking very clearly about what "required" means, which is narrowly defined as "field is set". I'll add additional validation around emptiness for strings and arrays.

**Comment by ironcladlou** [comment]
Latest commit:

1. Removes required tags from non-pointer structs
2. Adds min length validation to required strings
3. Adds min items validation to required arrays

File: operatoringress/v1/types.go [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	Targets []string `json:"targets"`
+	// recordType is the DNS record type. For example, "A" or "CNAME".
+	// +kubebuilder:validation:Required
+	// +required
+	RecordType DNSRecordType `json:"recordType"`
+	// recordTTL is the record TTL in seconds. If <= zero, the default is 30.
```
Comment (sttts) [inline operatoringress/v1/types.go]: If equal to zero

File: operatoringress/v1/types.go [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+	// +kubebuilder:validation:Minimum=0
+	// +required
+	RecordTTL int64 `json:"recordTTL"`
+}
+
+// DNSRecordStatus is the most recently observed status of each record.
+type DNSRecordStatus struct {
+	// zones are the status of the record in each zone.
+	//
+	// +kubebuilder:validation:Optional
```
Comment (sttts) [inline operatoringress/v1/types.go]: the default is optional. You only need a marker on required items.

File: operatoringress/v1/types.go (hunk 26 above)
Comment (ironcladlou) [reply operatoringress/v1/types.go]: Fixed

File: operatoringress/v1/types.go (hunk 27 above)
Comment (ironcladlou) [reply operatoringress/v1/types.go]: Fixed here and one other place
